File: Python/GA-VRP/Chrom.py
from math import fabs
from VRP import VRP
import random

class Chrom:
    def __init__(self, vrp) -> None:
        self.gene = []
        self.nodeList = vrp.nodeList.copy()
        self.uavList = vrp.uavList.copy()
        self.distance = vrp.dis.copy()
        self.nodeNum = len(self.nodeList)
        self.uavNum = len(self.uavList)
        self.mileage = [0] * self.uavNum
        self.load = [0] * self.uavNum
        self.valid = True
        self.k1 = vrp.k1
        self.k2 = vrp.k2
        self.k3 = vrp.k3
        self.time = 0
        self.length = 0
        self.cnt = 0

        for i in range(self.nodeNum - 1):   # 第一个点是配送中心
            self.gene.append(i + 1)

        # random.shuffle(self.gene)

        # index = 0
        # for i in range(self.uavNum - 1):
        #     demandSum = 0
        #     while True:
        #         if index == len(self.gene):
        #             break
        #         demandSum = demandSum + self.nodeList[self.gene[index]].demand
        #         if demandSum > self.uavList[i].capacity:
        #             break
        #         index += 1
        #     self.gene.insert(index, 0)
        self.gene.insert(2, 0)
        self.update()

    def update(self):
        for i in range(self.uavNum):
            self.mileage[i] = self.load[i] = 0
        uavIdx = 0
        last = 0
        for i in range(len(self.gene)):
            if self.gene[i] == 0:
                self.mileage[uavIdx] += self.distance[last][0]
                uavIdx += 1
                last = 0
            else:
                self.mileage[uavIdx] += self.distance[last][self.gene[i]]
                self.load[uavIdx] += self.nodeList[self.gene[i]].demand
                last = self.gene[i]
                if self.load[uavIdx] > self.uavList[uavIdx].capacity:
                    self.valid = False
                    return
        self.mileage[uavIdx] += self.distance[last][0]
        
        self.time = 0
        self.length = 0
        self.cnt = 0
        for mile in self.mileage:
            self.time = max(self.time, mile)
            self.length += mile
            if fabs(mile - 0.0) > 1e-6:
                self.cnt += 1
        self.valid = True

    def mutation(self):
        i = 0
        j = -1
        self.gene[0], self.gene[-1] = self.gene[-1], self.gene[0]
        t_time = self.time
        t_length = self.length
        t_cnt = self.cnt
        if self.valid == False:
            self.gene[0], self.gene[-1] = self.gene[-1], self.gene[0]
            self.time = t_time
            self.length = t_length
            self.cnt = t_cnt

        

    def fitness(self):
        if self.valid == False:
            return 1e8
        else:
            return self.k1 * self.time + self.k2 * self.length + self.k3 * self.cnt

    def infoNode(self):
        for node in self.nodeList:
            node.info()
    
    def infoGene(self):
        print(self.gene)
    

def main():
    vrp = VRP()
    vrp.addNode(15.58, 7.1, 0.9)
    vrp.addNode(10.26, 6.08, 1.7)
    vrp.addNode(17.54, 14.12, 0.4)
    vrp.addNode(16.92, 5.3, 2.7)
    vrp.addNode(15, 13.3, 2.8)
    # vrp.infoNode()
    vrp.addUAV(10, 20)
    vrp.addUAV(10, 20)
    vrp.solve()

    chrom = Chrom(vrp)
    chrom.infoNode()
    chrom.infoGene()

    chrom.mutation()
    chrom.infoGene()



if __name__ == '__main__':
    main()

File: C++/debug/VRP/Chrom.cpp
#include "Chrom.h"
#include "VRP.h"
#include "Random.h"

#include <algorithm>
#include <iostream>
#include <cmath>

using namespace std;

Chrom::Chrom(const VRP& vrp)
	: nodeInfo(vrp.nodeInfo), carInfo(vrp.carInfo), dis(vrp.dis)
{
	k1 = vrp.k1;
	k2 = vrp.k2;
	k3 = vrp.k3;

	int cNode = nodeInfo.size() - 1;
	int cCar = carInfo.size();
	
	mileage.resize(cCar, 0.0);
	load.resize(cCar, 0.0);
	
	for (int i = 0; i < cNode; ++i)
	{
		gene.push_back(i + 1);	// 初始化1~n
	}
	
	Random::Shuffle(gene);		// 打乱

	int index = 0;
	for (int i = 0; i < cCar - 1; ++i)	// 插入m-1个0
	{
		double sum = 0.0;
		while (1)
		{
			if (index >= (int)gene.size())	// 到了最后一个位置
			{
				break;
			}
			sum += nodeInfo[gene[index]].demand;
			if (sum > carInfo[i].capacity)	// 到这个位置刚好超过载荷限制
			{
				break;
			}
			index++;
		}

		gene.insert(gene.begin() + index, 0);
	}
	
	update();

	/*vector<int> p(cNode);
	for (int i = 0; i < cNode; ++i)
	{
		p[i] = i + 1;
	}
	Random::Shuffle(p);

	int iCar = 0, iNode = 0;
	double sum = 0.0;
	while (iCar < cCar - 1 && iNode < cNode)
	{
		if (sum > carInfo[iCar].capacity)
		{
			gene.push_back(0);
			iCar++;
			sum = 0.0;
		}
		else
		{
			gene.push_back(p[iNode]);
			sum += nodeInfo[p[iNode]].demand;
		}
		iNode++;
	}

	while (iCar < cCar - 1)
	{
		gene.push_back(0);
		iCar++;
	}

	update();*/
}

void Chrom::update()
{
	for (int i = 0; i < (int)mileage.size(); ++i)
	{
		mileage[i] = load[i] = 0.0;
	}

	int iCar = 0;
	int last = 0;
	for (int i = 0; i < (int)gene.size(); ++i)
	{
		if (gene[i] == 0)
		{
			mileage[iCar] += dis[last][0];
			iCar++;
			last = 0;
		}
		else
		{
			mileage[iCar] += dis[last][gene[i]];	// 每辆车的行驶距离
			load[iCar] += nodeInfo[gene[i]].demand;	// 每辆车的载荷
			last = gene[i];
			if (load[iCar] > carInfo[iCar].capacity)
			{
				valid = false;
				return;
			}
		}
	}
	mileage[iCar] += dis[last][0];	// 最后这辆车也要回到起点

	time = 0.0;
	length = 0.0;
	cnt = 0;
	for (int i = 0; i < (int)mileage.size(); ++i)
	{
		time = max(time, mileage[i]);	// 最长的路线对应最长的时间
		length += mileage[i];
		if (fabs(mileage[i] - 0.0) > 1e-6)	// 路线为0表示没用这辆车
		{
			cnt++;
		}
	}

	valid = true;
}

void Chrom::mutation()
{
	int i = Random::UniformInt(0, gene.size() - 1);
	int j = Random::UniformInt(0, gene.size() - 1);
	swap(gene[i], gene[j]);		// 随机交换两个基因
	double t_time = time;
	double t_length = length;
	int t_cnt = cnt;
	update();
	if (!valid)
	{
		swap(gene[i], gene[j]);
		time = t_time;
		length = t_length;
		cnt = t_cnt;
		valid = true;
	}
}

double Chrom::fitness() const
{
	if (!valid)
		return 1e8;
	else
		return k1 * time + k2 * length + k3 * cnt;
}

bool Chrom::operator<(const Chrom& c) const
{
	return fitness() < c.fitness();
}

Chrom& Chrom::operator=(const Chrom& c)
{
	gene = c.gene;
	mileage = c.mileage;
	load = c.load;
	time = c.time;
	length = c.length;
	cnt = c.cnt;
	valid = c.valid;
	return *this;
}

void Chrom::decode(Result& res) const
{
	res.path.resize(carInfo.size());
	res.load = load;
	res.mileage = mileage;

	int iCar = 0;
	for (int i = 0; i < (int)gene.size(); ++i)
	{
		if (gene[i] == 0)
		{
			iCar++;
		}
		else
		{
			res.path[iCar].push_back(gene[i]);
		}
	}
}

std::string Chrom::toString() const
{
	string s = "gene: ";
	for (int i = 0; i < (int)gene.size(); ++i)
	{
		s += to_string(gene[i]);
		s += " ";
	}
	return s;
}